### PR TITLE
[FEATURE] Enable `GCS DataConnector` integration with `PandasExecutionEngine`

### PR DESCRIPTION
**Pull request by cdkini**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- https://superconductive.atlassian.net/secure/RapidBoard.jspa?rapidView=24&projectKey=GREAT&modal=detail&selectedIssue=GREAT-158
- Allow `PandasExecutionEngine` to support `GCSDataConnectors` by being able to read the contents of GCS blobs
- Update existing unit tests and add new ones to gain coverage over the changes made herein
- Misc changes to documentation in files (primarily around Azure which I missed during my initial implementation)

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/execution_engine/pandas_execution_engine.py [hunk 1]
```diff
@@ -626,7 +682,7 @@ def _split_on_hashed_column(
         """Split on the hashed value of the named column"""
         try:
             hash_method = getattr(hashlib, hash_function_name)
-        except (TypeError, AttributeError) as e:
+        except (TypeError, AttributeError):
```
Comment (cdkini) [inline great_expectations/execution_engine/pandas_execution_engine.py:692]: We don't actually use the value of `e` so I've removed it from any applicable `except` clauses

File: great_expectations/core/util.py [hunk 2]
```diff
@@ -407,6 +407,7 @@ def datetime_to_int(dt: datetime.date) -> int:
 class AzureUrl:
     """
     Parses an Azure Blob Storage URL into its separate components
+    Format: <ACCOUNT_NAME>.blob.core.windows.net/<CONTAINER>/<BLOB>
```
Comment (cdkini) [inline great_expectations/core/util.py:410]: I've made a few Azure related changes here since I was already in the same files.

If you think I should move these to a separate PR, please let me know! Don't want to make reviewing more complicated than it needs to be

File: great_expectations/execution_engine/pandas_execution_engine.py [hunk 3]
```diff
@@ -16,27 +16,44 @@
...
     RuntimeDataBatchSpec,
     S3BatchSpec,
 )
-from great_expectations.core.util import AzureUrl, S3Url, sniff_s3_compression
+from great_expectations.core.util import AzureUrl, GCSUrl, S3Url, sniff_s3_compression
 from great_expectations.execution_engine import ExecutionEngine
 from great_expectations.execution_engine.execution_engine import MetricDomainTypes
 from great_expectations.execution_engine.pandas_batch_data import PandasBatchData
 
+logger = logging.getLogger(__name__)
```
Comment (cdkini) [inline great_expectations/execution_engine/pandas_execution_engine.py:30]: Moved logger up before importing cloud providers in order to use it for `logger.debug`

File: great_expectations/execution_engine/pandas_execution_engine.py [hunk 4]
```diff
@@ -94,17 +112,35 @@ def __init__(self, *args, **kwargs):
             self._s3 = None
 
         try:
-            self._azure = BlobServiceClient(**azure_options)
+            if "conn_str" in azure_options:
+                self._azure = BlobServiceClient.from_connection_string(**azure_options)
+            else:
+                self._azure = BlobServiceClient(**azure_options)
```
Comment (cdkini) [inline great_expectations/execution_engine/pandas_execution_engine.py:121]: Missed the alternative constructor on my first implementation of this! These `try/except` clauses should mirror what happens in the `DataConnector` constructors

File: great_expectations/execution_engine/pandas_execution_engine.py [hunk 5]
```diff
@@ -201,6 +237,26 @@ def get_batch_data_and_markers(
...
+            gcs_bucket = gcs_engine.get_bucket(gcs_url.bucket)
+            gcs_blob = gcs_bucket.blob(gcs_url.blob)
+            logger.debug(
+                f"Fetching GCS blob. Bucket: {gcs_url.bucket} Blob: {gcs_url.blob}"
+            )
+            reader_fn = self._get_reader_fn(reader_method, gcs_url.blob)
+            buf = BytesIO(gcs_blob.download_as_bytes())
+            buf.seek(0)
+            df = reader_fn(buf, **reader_options)
+
```
Comment (cdkini) [inline great_expectations/execution_engine/pandas_execution_engine.py:266]: Identical to the processing done for `Azure` and `S3` with slight deviations in the specific SDK methods use to get the appropriate bucket/blob

File: requirements-dev-base.txt [hunk 6]
```diff
@@ -16,6 +16,7 @@ feather-format>=0.4.1  # all_tests
 flake8==3.8.3  # lint
 flask>=1.0.0 # for s3 test only
 freezegun>=0.3.15  # all_tests
+gcsfs>=0.5.1 # all_tests
```
Comment (cdkini) [inline requirements-dev-base.txt:19]: The Google SDK has an implicit subdependency on `gcsfs` to actually read blob contents using the `gs://` URL format. I couldn't find it noted anywhere in the documentation but I ran into errors without it.

Let's keep tabs on this one as I don't have a lot of faith in these `fs` libs after that whole Azure fiasco

File: tests/core/test_util.py [hunk 7]
```diff
@@ -83,3 +84,15 @@ def test_azure_url_with_nested_blob():
...
+def test_gcs_url():
+    url = GCSUrl("gs://my_bucket/my_blob")
+    assert url.bucket == "my_bucket"
+    assert url.blob == "my_blob"
+
+
+def test_gcs_url_with_nested_blob():
+    url = GCSUrl("gs://my_bucket/a/b/c/d/e/my_blob")
+    assert url.bucket == "my_bucket"
+    assert url.blob == "a/b/c/d/e/my_blob"
```
Comment (cdkini) [inline tests/core/test_util.py:114]: Very simple but like my Azure implementation, I'm always wary of blowing something up with RegEx regardless of how trivial the problem might be 💥

File: tests/execution_engine/test_pandas_execution_engine.py [hunk 8]
```diff
@@ -939,7 +962,7 @@ def test_get_batch_with_split_on_divided_integer_and_sample_on_list(test_df):
 
 
 @mock.patch(
-    "great_expectations.datasource.data_connector.configured_asset_azure_data_connector.BlobServiceClient"
+    "great_expectations.execution_engine.pandas_execution_engine.BlobServiceClient",
```
Comment (cdkini) [inline tests/execution_engine/test_pandas_execution_engine.py:965]: A mock should always reference **WHERE** it is invoked as opposed to where it comes from. This was erroneous but still worked before?

Just changing to be consistent

File: tests/data_context/test_data_context_test_yaml_config.py [hunk 9]
```diff
@@ -1159,6 +1159,7 @@ def test_golden_path_runtime_data_connector_pandas_datasource_configuration(
         "discard_subset_failing_expectations": False,
         "boto3_options": {},
         "azure_options": {},
+        "gcs_options": {},
```
Comment (cdkini) [inline tests/data_context/test_data_context_test_yaml_config.py:1162]: Small change just to get assertions passing for Data Context tests

File: great_expectations/core/util.py [hunk 10]
```diff
@@ -430,6 +431,26 @@ def blob(self):
...
+class GCSUrl:
+    """
+    Parses a Google Cloud Storage URL into its separate components
+    Format: gs://<BUCKET_OR_NAME>/<BLOB>
+    """
+
+    def __init__(self, url: str):
+        search = re.search(r"^gs://([^/]+)/(.+)$", url)
+        self._bucket = search.group(1)
+        self._blob = search.group(2)
```
Comment (NathanFarmer) [inline great_expectations/core/util.py]: Are these values used elsewhere and if so is a `try`/`except` being used?

File: tests/core/test_util.py (hunk 7 above)
Comment (NathanFarmer) [reply tests/core/test_util.py:114]: This was the reason for my comment above. Would it make sense to try some special characters and/or a negative assertion condition here?

**Review by NathanFarmer — COMMENTED** [review]
See my comments inside, otherwise LGTM.

File: great_expectations/core/util.py (hunk 10 above)
Comment (cdkini) [reply great_expectations/core/util.py]: The URL passed here directly comes from the `GCSDataConnector`'s `_get_full_file_path()` method:

```
    def _get_full_file_path(
        self,
        path: str,
        data_asset_name: Optional[str] = None,
    ) -> str:
        # data_asset_name isn't used in this method.
        # It's only kept for compatibility with parent methods.
        return f"gs://{os.path.join(self._bucket_or_name, path)}"
```

as such, it must conform the this `gs://<BUCKET_OR_NAME>/<BLOB>` pattern. The values parsed here (`bucket` and `blob`) are used in the `PandasExecutionEngine` to find a specific blob and read it into memory. You can find that usage in L249 of the `PandasExecutionEngine` in this PR.

File: tests/core/test_util.py (hunk 7 above)
Comment (cdkini) [reply tests/core/test_util.py:114]: The URL should always be in the `gs://<BUCKET>/<BLOB>` format as noted by the GCS SDK but I don't think there's any harm in adding additional tests around the two conditions you noted!

Will do that shortly. Thanks!

File: great_expectations/core/util.py (hunk 10 above)
Comment (cdkini) [reply great_expectations/core/util.py]: We don't use `try/except` anywhere. Do you think it is appropriate to place in the `PandasExecutionEngine`? Or perhaps even a simple `assert` after `search` to ensure that `re.search` actually parses correctly? The assertion should never be triggered but I don't think it hurts to add a check to see if we violate the invariant.

Comment (NathanFarmer) [reply great_expectations/core/util.py]: Makes perfect sense, thanks.

Comment (NathanFarmer) [reply great_expectations/core/util.py]: @cdkini I missed your second comment above. I don't have a deep understanding of GCS or all the ways you can interact with the connector, but if the only source for this string is from the GCS filesystem then your assumptions are pretty solid. The only risk would be if GCS introduces a breaking change to that pattern. If the user can pass a bad string it will still fail somewhere, but they may not get a very useful error message. These diagrams may be helpful: [https://docs.greatexpectations.io/docs/reference/datasources](https://docs.greatexpectations.io/docs/reference/datasources).

Comment (cdkini) [reply great_expectations/core/util.py]: from my understanding, the only way our system uses a `GCS` URL is from that `_get_full_file_path` method from the `DataConnector`. I think the assertion is helpful, especially since we're using regular expressions.

I'll look into this reference just to confirm my understanding. Thanks Nathan!

File: great_expectations/core/util.py [hunk 11]
```diff
@@ -430,6 +434,29 @@ def blob(self):
...
+    """
+    Parses a Google Cloud Storage URL into its separate components
+    Format: gs://<BUCKET_OR_NAME>/<BLOB>
+    """
+
+    def __init__(self, url: str):
+        search = re.search(r"^gs://([^/]+)/(.+)$", url)
+        assert (
+            search is not None
+        ), "The provided URL does not adhere to the format specified by the GCS SDK (gs://<BUCKET_OR_NAME>/<BLOB>)"
```
Comment (cdkini) [inline great_expectations/core/util.py:447]: Invariant that should theoretically never be violated but a good practice given our usage of regular expressions

File: great_expectations/execution_engine/pandas_execution_engine.py [hunk 12]
```diff
@@ -94,17 +115,39 @@ def __init__(self, *args, **kwargs):
...
+                    credentials = service_account.Credentials.from_service_account_file(
+                        **gcs_options
+                    )
+                elif "info" in gcs_options:
+                    credentials = service_account.Credentials.from_service_account_info(
+                        **gcs_options
+                    )
+                self._gcs = storage.Client(credentials=credentials, **gcs_options)
+            except (TypeError, AttributeError):
+                self._gcs = None
```
Comment (cdkini) [inline great_expectations/execution_engine/pandas_execution_engine.py:141]: An improper instantiation of `storage.Client` will cause a warning to be raised by the Google SDK. This is not a big deal but it breaks hundreds of tests in our `test_cli` Azure stage.

I added a simple check for either environment variables or explicit options to possibly set `self._gcs` to `None` before even going through that process.

File: great_expectations/core/util.py (hunk 2 above)
Comment (Shinnnyshinshin) [reply great_expectations/core/util.py:410]: It's ok. The changes are small and are pretty closely related with GCSUrl

File: great_expectations/core/util.py (hunk 11 above)
Comment (Shinnnyshinshin) [reply great_expectations/core/util.py:447]: Thank you for this :)

File: great_expectations/execution_engine/pandas_execution_engine.py (hunk 12 above)
Comment (Shinnnyshinshin) [reply great_expectations/execution_engine/pandas_execution_engine.py:141]: This makes sense :)

File: great_expectations/execution_engine/pandas_execution_engine.py (hunk 1 above)
Comment (Shinnnyshinshin) [reply great_expectations/execution_engine/pandas_execution_engine.py:692]: ❤️

**Review by Shinnnyshinshin — APPROVED** [review]
Looks great :) Thank you
